lib/src/test/java/com/basistheory/android/service/HTTPClientTests.kt

```diff
@@ -0,0 +1,200 @@
+        assertEquals("PATCH", request.method)
+        assertEquals(endpoint, request.path)
+        assertEquals(expectedRequestBody, request.body.readUtf8())
+    }
+
+    @Test
+    fun testDeleteRequest() = runBlocking {
+        server.enqueue(MockResponse().setBody("Deleted it"))
+
+        val url = server.url(endpoint).toString()
+
+        client.delete(url)
+
+        val request = server.takeRequest()
+
+        assertEquals("DELETE", request.method)
+        assertEquals(endpoint, request.path)
+    }
+
+
+    @Test
+    fun testFormUrlEncodedPostRequest() = runBlocking {
+        val expectedRequestBody = object {
+            val type = "card"
+            val billing_details = object {
+                val name = "Peter Panda"
+            }
+            val card = object {
+                val number = "1234567890"
+                val exp_month = 12
+                val exp_year = 2023
+                val cvc = 123
+            }
+        }
+
+        server.enqueue(MockResponse().setBody("Encoded it"))
+
+        val url = server.url(endpoint).toString()
+
+        client.post(
+            url,
+            expectedRequestBody,
+            headers = mapOf("Content-Type" to "application/x-www-form-urlencoded")
+        )
+
+        val request = server.takeRequest()
+        assertEquals("POST", request.method)
+        assertEquals(endpoint, request.path)
+        assertEquals(
+            "type=card&billing_details%5Bname%5D=Peter+Panda&card%5Bnumber%5D=1234567890&card%5Bexp_month%5D=12&card%5Bexp_year%5D=2023&card%5Bcvc%5D=123",
+            request.body.readUtf8()
+        )
+
+    }
+
+    @Test
+    fun testDefaultsContentTypeToApplicationJsonWhenNoContentTypeIsProvided() = runBlocking {
+        val reqBody = object {
+            val foo = "bar"
+        }
+
+        server.enqueue(MockResponse().setBody("Defaulted it"))
+
+        val url = server.url(endpoint).toString()
+
+        client.post(
+            url,
+            reqBody,
+        )
+
+        val request = server.takeRequest()
+
+        assert(request.headers["Content-Type"]!!.contains("application/json"))
+    }
+
+
+    @Test
+    fun `testHandlesUnsuccessfulResponsesFromTheRemote`() = runBlocking {
+        val reqBody = object {
+            val foo = "bar"
+        }
+
+        server.enqueue(MockResponse().setBody("Everything is broken").setResponseCode(500))
+
+        val url = server.url(endpoint).toString()
+
+        val exception = assertThrows(IOException::class.java) {
+           runBlocking {
+               client.post(url, reqBody) as IOException
+           }
+        }
+
+        assertEquals("POST request failed with response code: 500", exception.message)
+    }
+
+    @Test
+    fun `testThrowsWhenContentTypeIsNotSupported`() = runBlocking {
+        val reqBody = object {
+            val foo = "bar"
+        }
+
+        server.enqueue(MockResponse().setBody("Everything is broken").setResponseCode(500))
```

Comment:
does it really returns 500? it feels like we should return 400

Reply:
This is just MockedWebServer; it could return anything you want and in this test we're just making sure we throw when using unsupported `content-type`s in the request; we don't have control over the responses from the 3rd-parties anyways

Reply:
but if they provide a content type that the client doesn't support does it return 500?